### PR TITLE
Fix #6776 - Add new custom targeting for Fx95 users

### LINKED ISSUE #6776: Add new custom targeting for Fx 95 users only
Till #6775 is resolved, we need custome targeting to target 95 users only

`version|versionCompare('95.!') == 0`

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1939)

### PR DESCRIPTION
**Pull request by punamdahiya**

### CODE REVIEW

**Review by jaredlockhart — APPROVED** [review]
LGTM 👍

File: app/experimenter/experiments/constants/nimbus.py [hunk 1]
```diff
@@ -213,6 +213,13 @@ class NimbusTargetingConfig:
     application_choice_names=(Application.DESKTOP.name,),
 )
 
+TARGETING_FX95_DESKTOP_USERS = NimbusTargetingConfig(
+    name="Desktop Users on Fx95",
+    slug="fx95_desktop_users",
+    description=("Firefox 95 Desktop users"),
+    targeting=("version|versionCompare('95.!') == 0"),
```
Comment (Mardak) [inline app/experimenter/experiments/constants/nimbus.py:220]: This matches exactly the version "95.!" where I believe "!" is used as a "dummy" character ("less than" `0`) to relatively compare against. If we want any 95 (e.g., including "a" and "b" and dot releases), we probably want a range, e.g., compare(95) >= 0 && compare(96) < 0

**Review by Mardak — CHANGES_REQUESTED** [review]
https://searchfox.org/mozilla-central/source/xpcom/base/nsIVersionComparator.idl

File: app/experimenter/experiments/constants/nimbus.py (hunk 1 above)
Comment (punamdahiya) [reply app/experimenter/experiments/constants/nimbus.py:220]: +1, should the range be "version|versionCompare('95') >= 0 && version|versionCompare('96') < 0"

Or with dummy character  "version|versionCompare('95.!’) >= 0 && version|versionCompare('96.!’) < 0" or either works

Comment (jaredlockhart) [reply app/experimenter/experiments/constants/nimbus.py:220]: Oh you're right, that doesn't work.  On my local:

```js
await targetingContext.evalWithDefault("version")
"96.0b4"

await targetingContext.evalWithDefault("version|versionCompare('95.!') == 0")
false

await targetingContext.evalWithDefault("version|versionCompare('96.!') == 0")
false

await targetingContext.evalWithDefault("version|versionCompare('97.!') == 0")
false
```

Using the method @Mardak is suggesting:

```js
await targetingContext.evalWithDefault("(version|versionCompare('95.!') >= 0) && (version|versionCompare('96.!') < 0)")
false

await targetingContext.evalWithDefault("(version|versionCompare('96.!') >= 0) && (version|versionCompare('97.!') < 0)")
true

await targetingContext.evalWithDefault("(version|versionCompare('97.!') >= 0) && (version|versionCompare('98.!') < 0)")
false
```

**Review by jaredlockhart — CHANGES_REQUESTED** [review]
Actually I think this doesn't work as intended, see my comment above

File: app/experimenter/experiments/constants/nimbus.py (hunk 1 above)
Comment (punamdahiya) [reply app/experimenter/experiments/constants/nimbus.py]: thanks for verifying,  patch updated with feedback

File: app/experimenter/experiments/constants/nimbus.py [hunk 2]
```diff
@@ -213,6 +213,15 @@ class NimbusTargetingConfig:
     application_choice_names=(Application.DESKTOP.name,),
 )
 
+TARGETING_FX95_DESKTOP_USERS = NimbusTargetingConfig(
+    name="Desktop Users on Fx95",
+    slug="fx95_desktop_users",
+    description=("Firefox 95 Desktop users"),
+    targeting=("(version|versionCompare('95.!') >= 0) && (version|versionCompare('96.!') < 0)"),
```
Comment (Mardak) [inline app/experimenter/experiments/constants/nimbus.py:220]: Pretty sure operator precedence will check the comparisons before the AND, but this is fine.

FYI (i.e., don't update patch) If you really wanted to "optimize" 😝  and maximize the mozilla version comparison quirks.. `version|versionCompare('95+') < 0` works for the second part and `version|versionCompare('95pre') >= 0` for the first………

**Review by Mardak — APPROVED** [review]
Just making sure, you don't need any other "advanced targeting" for the experiment (i.e., it's "No Targeting")? If not, you'll need to combine the two advanced targeting as only 1 can be selected.

**Comment by punamdahiya** [comment]
+1, yes the experiment was not using any advanced targeting. With this custom targeting available , we can set Version drop down to 'No version' and pick 'TARGETING_FX95_DESKTOP_USERS' from Advanced Targeting

**Comment by Mardak** [comment]
> we can set Version drop down to 'No version'

I suppose technically we could have just exposed this as "max 95 version" and reused the existing minimum version targeting if for some reason we need to target say Fx94 & Fx95 for some reason for another experiment. I'm not familiar with how these one-off advanced targeting should get cleaned up -- especially with this particular targeting is probably less likely to be reused.

File: app/experimenter/experiments/constants/nimbus.py (hunk 2 above)
Comment (jaredlockhart) [reply app/experimenter/experiments/constants/nimbus.py:220]: Yeah I would've thought so too but it didn't work in my local testing without them

Comment (jaredlockhart) [reply app/experimenter/experiments/constants/nimbus.py:220]: Also interesting good to know 👍

**Comment by punamdahiya** [comment]
Eventually we should get capability to set Max version from console (#6775). Just left comment in #6775 to remove this custom targeting with its fix

**Review by jaredlockhart — APPROVED** [review]

**Comment by jaredlockhart** [comment]
Hrmmm something's gone wrong with the tests that validate the targeting in Firefox.  I'll have to poke at it some more in the morning.

**Comment by punamdahiya** [comment]
hmm why are the tests failing for , where it should skip?
[gw0] [ 87%] FAILED nimbus/test_remote_settings.py::test_check_targeting[IOS-fx95_desktop_users] 
nimbus/test_remote_settings.py::test_check_targeting[KLAR-fx95_desktop_users] 
[gw1] [ 91%] FAILED nimbus/test_remote_settings.py::test_check_targeting[FENIX-fx95_desktop_users] 
nimbus/test_remote_settings.py::test_check_targeting[FOCUS-fx95_desktop_users] 
[gw0] [ 95%] FAILED nimbus/test_remote_settings.py::test_check_targeting[KLAR-fx95_desktop_users] 
[gw1] [100%] FAILED nimbus/test_remote_settings.py::test_check_targeting[FOCUS-fx95_desktop_users]

**Comment by jaredlockhart** [comment]
@punamdahiya yeah it's a weird edge case in our pytest setup, I added a hack to skip those tests (which shouldn't be running in the first place) and filed a followup https://github.com/mozilla/experimenter/issues/6791  The incomparable @jrbenny35 is gonna look into it.

File: app/tests/integration/nimbus/test_remote_settings.py [hunk 3]
```diff
@@ -7,6 +7,7 @@
 from nimbus.pages.experimenter.summary import SummaryPage
 from nimbus.pages.remote_settings.dashboard import Dashboard
 from nimbus.pages.remote_settings.login import Login
+from nimbus.models.base_dataclass import BaseExperimentApplications
```
Comment (punamdahiya) [inline app/tests/integration/nimbus/test_remote_settings.py]: @jaredlockhart looks like test failed bcause of import order 

ERROR: /app/tests/integration/nimbus/test_remote_settings.py Imports are incorrectly sorted and/or formatted.

**Comment by punamdahiya** [comment]
Thank you!
